### PR TITLE
PCSX2: Fix stall on branch .. in delay slot

### PR DESCRIPTION
**Pull request by FlatOutPS2**
Fixes stall when loading a stage in WRC 3.

Removes patch that served as workaround.

### CODE REVIEW

**Comment by refractionpcsx2** [comment]
That was quick! I assumed branches in the delay slot would be handled like the vu ones, which would be a pig to fix. 

Trying to think if there's any other games that do this too. 

Good job

**Comment by refractionpcsx2** [comment]
Apparently wrc really evolved and possibly midnight club 2 suffered from these too

**Comment by MonJamp** [comment]
@FlatOutPS2 Do you know if this fixes enabling MTVU for Thrillville (you have the game right)?

**Comment by refractionpcsx2** [comment]
This won't fix games that break when hacks are enabled. The mtvu issue is completely separate

**Comment by willkuer** [comment]
May I ask how you came to this solution? Were you checking what the hack did? I guess there is no forum in the web where people discussed 'well, if you have a stall in branch … delay slot, just put these lines and the errors are gone…' and ps2 documentation also rather seems poor...

**Comment by refractionpcsx2** [comment]
PS2 documentation is poor and I'm not sure how the EE handles it.

On the VU side, what happens is the VU will take the first branch, read the first instruction, then take the second branch (if it succeeded).  Of course if the first branch wasn't taken, as long as it isn't a Likely branch (speaking from the EE side, the VU doesn't have likely's), it would take the second branch only.

however it seems there is an exception for when there is a branch in the branch delay slot, so it's possible the EE does handle it completely differently and just hands it off to the exception manager to deal with.

**Comment by gregory38** [comment]
I would suggest to run the EE PS2 testsuite. Initial code got an infinite function call loop.
The global variables to handle the flush are not really used. (All registers allocation is completely broken, the big part to redo on x64)

**Comment by ramapcsx2** [comment]
Woho, a EE recompiler fix. Awesome!
Would be great to test this on a few more games.
The positive (buggy) case will be very clear: It writes "branch xxxx in delay slot!" :)

**Comment by prafullpcsx2** [comment]
This looks interesting. Will test it thoroughly after my tournament ends in two days time.

File: pcsx2/x86/ix86-32/iR5900-32.cpp [hunk 1]
```diff
@@ -1315,17 +1315,29 @@ void recompileNextInstruction(int delayslot)
...
 					case 0: case 1: case 2: case 3: case 0x10: case 0x11: case 0x12: case 0x13:
-						Console.Warning("branch %x in delay slot!", cpuRegs.code);
+						DevCon.Warning("branch %x in delay slot!", cpuRegs.code);
 						_clearNeededX86regs();
 						_clearNeededXMMregs();
+						pc += 4;
+						g_cpuFlushedPC = false;
+						g_cpuFlushedCode = false;
+						if (g_maySignalException)
+							xAND(ptr32[&cpuRegs.CP0.n.Cause], ~(1 << 31)); // BD
```
Comment (ssakash) [inline pcsx2/x86/ix86-32/iR5900-32.cpp]: According to the documents, BD consists of two bits (BD & BD2) and both of them seem to have the same functionality, shouldn't you also clear the second bit if you're doing it for the first one? Any specific reason to only clear the first one?

**Comment by ssakash** [comment]
> That said we aren't actually setting the BD bit, we're just clearing it?

Yeah, doesn't make sense to me either. According to the sources, we only need to set BD when encountering exceptions other than the reset, NMI, performance counter and debug on the delay slot. I don't seem to find any information about clearing the bit at any cases.

Also seems like @sudonim1 was the one who introduced it in https://github.com/PCSX2/pcsx2/commit/d4bc1a62755b6d4fcaf873d85bb03ef61b3d5e0a, so assigning this to him as he probably knows how this works.

**Comment by gregory38** [comment]
@ssakash here the theory
When you encounter an exception/interrupt, you need to stop the program to handle the exception/interrupt. When it is done you must resume the execution. Therefore you must save the PC before jumping to the exception handler (saved in the EPC register)
For example:
```
0x400:load v0, 0;
0x404:load v1, v0; // may generate a TLB miss
0x408:load v0, v0;
```
2nd load will generate a TLB miss. So you save the PC (here 0x404) in the EPC register. You execute the TLB miss handler. Then you return to the program with a `jump $EPC`.

Now imagine, that exception occurred in a delay slot.
```
0x400:load v0, 0;
0x404:jump 0x410;
0x408:load v1, v0; // may generate a TLB miss
0x40C:load v0, 1;
0x410: print $v0;
```
Again the CPU will save the PC of the instruction that raise the exception. Here 0x408. Handler will be executed. And finally jump $EPC (reminder 0x408). So next instruction will be 0x40C and 0x410. As you can see you didn't resume the execution correctly (will print 1). An exception/interrupt must never change the program behavior. So instead you save $EPC + the BD flag. This way, you will jump back to $EPC-4 (so 0x404) so program will resume correctly (will print 0).

There is 2 level of exceptions, hence BD/BD2. It is easier to handle in the CPU (and potentially it allow lvl2 exception in lvl1 exception). LVL2 exception must never happen in the emulation.

Now remain to answer the behavior of  what must happen if there is a branch in the delay slot case. I don't remember the exact behavior. Either 2nd jump is a NOP or final address is the one of the 2nd jump. Behavior can be checked with the testsuite.
```
load v0, imm;
jump 0x1000;
jump 0x2000; // jump in delay slot (could trigger a tlb miss)
jump 0x3000;
print PC; <= likely 0x1000 (or 0x2000)
```

BD will be cleared when you resume program execution (check ERET opcode). Potentially the behavior is equivalent to (for every cycle)
```
if (cpu is in std mode)
  bd = jump/branch opcode ? 1 : 0;
else if (cpu is in exception 1 mode)
  bd2 = jump/branch opcode ? 1 : 0;
else
   nop
```

**Comment by gregory38** [comment]
@FlatOutPS2 
I forgot to say in my wall of text, could you create an handy function to avoid copy/past.

**Comment by ssakash** [comment]
@gregory38 

Thanks for the write up, really appreciated. :)

> I forgot to say in my wall of text, could you create an handy function to avoid copy/past.

Agreed, I was also planning on mentioning it but forgot about it for some reasons.

**Comment by sudonim1** [comment]
I think that commit of mine was just preparation for future work that has only kind of happened because at least at the time it was impossible for any recompiler instruction to raise an exception.

**Comment by gregory38** [comment]
Please add a common function. And I need to run the test suite.

**Comment by gregory38** [comment]
do you have the instructions around the branch in branch delay slot.

From the HW testsuite:
```	int result = -1;

	// This is technically not valid MIPS anyway.
	// But it seems to always take target 2?
	asm volatile (
		".set    noreorder\n"

		"li      $t0, 0\n"
		"beq     $t0, $t0, target1_%=\n"
		"beq     $t0, $t0, target2_%=\n"
		"beq     $t0, $t0, target3_%=\n"
		"beq     $t0, $t0, target4_%=\n"
		"nop\n"

		"target1_%=:\n"
		"li      %0, 1\n"
		"j       skip_%=\n"
		"nop\n"

		"target2_%=:\n"
		"li      %0, 2\n"
		"j       skip_%=\n"
		"nop\n"

		"target3_%=:\n"
		"li      %0, 3\n"
		"j       skip_%=\n"
		"nop\n"

		"target4_%=:\n"
		"li      %0, 4\n"
		"j       skip_%=\n"
		"nop\n"

		"skip_%=:\n"

		: "+r"(result)
	);

	printf("beq: delay branch: %08x\n", result);
```
So far we return 1 (even with this PR), and the HW 2.

**Comment by sudonim1** [comment]
So basically branch in branch delay slot on real hardware works like you'd
expect.  The delay slot is fetched/the first branch executed, the first
branch target fetched/the second branch executed, the first branch target
executed/the second branch target fetched, execution continues from the
second branch target.

**Comment by gregory38** [comment]
@FlatOutPS2 could you create a generic function ?

**Comment by FlatOutPS2** [comment]
> could you create a generic function ?

I already did, but I was still to look into some possibly related issues before updating the PR. I just haven't gotten around to it yet.

**Comment by atomic83GitHub** [comment]
Any status?

**Comment by MrCK1** [comment]
I don't think this will be merged anytime soon. Flat out was in charge and we'll, he's gone for now.

**Comment by atomic83GitHub** [comment]
Apparently Midnight club games are fixed with this PR.

**Comment by atomic83GitHub** [comment]
I just seen that on a forum post (about the 2nd if a remember).

**Comment by FlatOutPS2** [comment]
Someone once said != actually true.

I don't see any post on the forum about testing this PR with that game.

**Comment by atomic83GitHub** [comment]
This wasn't in the PCSX2 forum, I unfortunately don't remember it because It's been a long time I seen that :). Anyway is there any other games affected by this?

**Comment by FlatOutPS2** [comment]
> Anyway is there any other games affected by this?

No. The only other mentions of this issue are due to cheats/hacks.

**Comment by refractionpcsx2** [comment]
The only other game I know I mentioned earlier in the thread is WRC 4 i think it is

**Comment by FlatOutPS2** [comment]
WRC Evolved. ;) Though it should also WRC 4, but that has other issues not related to this PR.

**Comment by atomic83GitHub** [comment]
The Action Replay MAX throw this in the console when applying Ratchet and Clank 2 cheats (which instantly close PCSX2):

capture

But even with your PR, it won't work.

**Comment by atomic83GitHub** [comment]
I tested this PR with a lot of games and I have 2 improvement:

-Action replay MAX
-WRC 2005 avec Sébastien Loeb

I think it's safe to merge as I did not notice any regression with it.

**Comment by lightningterror** [comment]
@refractionpcsx2 , @ramapcsx2 what are your thoughts, what are we gonna do with this?
@atomic83GitHub seems to have tested a bunch of games with no new issues popping up.

**Comment by refractionpcsx2** [comment]
If it works, merge it. Better than what we've got. however due tot he gamedb conflicts you might have to open a new PR

**Comment by ramapcsx2** [comment]
@lightningterror 
If you're doing a new PR, could you add a comment somewhere that links to the earlier discussion here?
It seems this code change works in favor of the games, but is still not entirely correct.
@gregory38 tested this in 2017 using the ps2autotests suite and remarked "So far we return 1 (even with this PR), and the HW 2."

I feel this bit shouldn't be lost. A comment could either link here, or summarize that there's still problems with the test suite.
Thanks :)

**Comment by lightningterror** [comment]
Closing in favor of https://github.com/PCSX2/pcsx2/pull/2919

**Pull request closed without merging**